File: Python/49 Group Anagram.py
def groupAnagram(strs: list[str]) -> list[list[str]]:
    storage = {}
    for value in strs:
        char_storage = [0] * 26
        for char in value:
            char_storage[ord(char) - ord('a')] += 1

        if tuple(char_storage) in storage:
            storage[tuple(char_storage)].append(value)
        else:
            storage[tuple(char_storage)] = [value]

    return list(storage.values())


print(groupAnagram(["eat", "tea", "tan", "ate", "nat", "bat"]))

File: TypeScript/49. Group Anagram.ts
function groupAnagram(strs: string[]): string[][]{
    interface storageType {
        [key: string]: string[];
    }
    const storage: storageType = {};
    for(const str of strs){
        const char_storage: number[] = new Array(26).fill(0);
        for(const char of str) char_storage[char.charCodeAt(0) - 97]++;
        const key = char_storage.join("");
        storage[key] ? storage[key].push(str) : storage[key] = [str];
    }
    return Object.values(storage);
}

const strs: string[] = ["eat", "tea", "tan", "ate", "nat", "bat"];
console.log(groupAnagram(strs));